File: 2022/go/7.go
package main

import (
	"fmt"
	"os"
	"strings"
	"strconv"
)

func min(a, b uint32) uint32 {
	if a < b {
		return a
	}
	return b
}

func main() {

	input, err := os.ReadFile("../7.txt")
	if err != nil {
		fmt.Printf("Could not read the file due to this %s error \n", err)
	}
	sc := string(input)
	lines := strings.Split(sc, "\n")

	paths := map[string]uint32{}

	path := []string{}
	for i := 0; i < len(lines); i++ {
		parsed_line := strings.Split(lines[i], " ")

		if parsed_line[1] == "cd" {
            if parsed_line[2] == ".." {
                path = path[:len(path)- 1]
            } else {
                path = append(path, parsed_line[2])
            }
        } else if parsed_line[1] == "ls" {
            continue
        } else if parsed_line[0] == "dir" {
            continue
        } else {
			conv_size,_ := strconv.ParseUint(parsed_line[0], 10, 32)
			size := uint32(conv_size)
			cur_path := ""
			for j := 0; j < len(path); j++ {
				cur_path += path[j]
				paths[cur_path] += size		
            }
        }     
	}


	var score_part1 uint32 = 0
	var score_part2 uint32 = 4294967295

	var limit uint32 = 100000
    var max uint32 = uint32(70000000) - uint32(30000000)
    var used uint32 = paths["/"]
    var need_to_free uint32 = used - max
	fmt.Printf("need_to_free: %v\n", need_to_free)	
	fmt.Printf("used: %v\n", used)	
	fmt.Printf("max: %v\n", max)	


	for _, val := range paths {
        if val <= limit {
			score_part1 += val
		} 
		if val >= need_to_free {
			score_part2 = min(val, score_part2)
		}
    }
	
	fmt.Printf("Score 1: %v\n", score_part1)	
	fmt.Printf("Score 2: %v\n", score_part2)
}

File: 2022/rust/7.rs
use std::fs;
use std::collections::HashMap;
use std::path::PathBuf;
use std::path::Path;
use std::iter::FromIterator;
use std::cmp;

fn main() {
    let input = fs::read_to_string("../7.txt").unwrap();
    let lines : Vec<&str> = input.split("\n").collect();
    let mut paths = HashMap::new();

    let mut path  = Vec::new();

    for line in &lines {
        let parsed_line : Vec<&str> = line.trim().split_whitespace().collect();

        if parsed_line[1] == "cd" {
            if parsed_line[2] == ".." {
                path.pop();
            }
            else {
                path.push(parsed_line[2]);
            }
        }
        else if parsed_line[1] == "ls" {
            continue;
        }
        else if parsed_line[0] == "dir" {
            continue;
        }
        else {

            let size = parsed_line[0].parse::<u32>().unwrap();
            let n = path.len();
            
            for i in 0..n {
                let cur_path = PathBuf::from_iter(&path[..=i]);
                *paths.entry(cur_path).or_insert(0) += size;
            }
        }     
    }

    let mut score_part1 : u32 = 0;
    let mut score_part2 : u32 = 4294967295;
    let limit : u32 = 100000;
    let max : u32 = 70000000 - 30000000;
    let used : &u32 = paths.get(&Path::new("/").to_path_buf()).unwrap();
    let need_to_free : u32 = used - max;
    println!("Need to free: {}", need_to_free);

    for (key, val) in paths.iter() {
        if val <= &limit {
            score_part1 += val;
        }
        if val >= &need_to_free {
            score_part2 = cmp::min(val.clone(), score_part2);
        }
    }

    println!("Score 1: {}", score_part1);
    println!("Score 2: {}", score_part2);
}